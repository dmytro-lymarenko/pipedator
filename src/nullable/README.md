- **Description**

Creates a validator which allows a value to be null in addition to the provided validator.

- **How to import**

```typescript
import { nullable } from 'pipedator';
// or
import { nullable } from 'pipedator/lib/nullable';

```
- **Signature**

```typescript
function nullable<ValidValue = any>(validator: Validator, message?: string): Validator<ValidValue>;
```
- **Parameters**

  - `validator` - some valildator to validate a value.
  - `message` - (optional) custom message.


- **Usage**

```typescript
const optionalNumber = nullable(number());
optionalNumber.validate(4); // valid
optionalNumber.validate(null); // valid
optionalNumber.validate(undefined); // invalid
```