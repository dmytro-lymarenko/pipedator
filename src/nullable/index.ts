import { Validator } from '../core';
import { either } from '../either';
import { equalTo } from '../equalTo';

export function nullable<ValidValue = any>(validator: Validator, message?: string) {
	return either<ValidValue>([equalTo(null), validator], message);
}
