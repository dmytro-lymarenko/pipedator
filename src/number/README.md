- **Description**

Creates a validator to check if a value is a number or not.

- **How to import**

```typescript
import { number } from 'pipedator';
// or
import { number } from 'pipedator/lib/number';

```
- **Signature**

```typescript
function number(message?: string): Validator;
```
- **Parameters**

  - `message` - (optional) custom message.


- **Usage**

```typescript
number().validate(5); // valid
number().validate(NaN); // valid
number().validate('Some text'); // invalid
number('Custom message').validate(null); // invalid with 'Custom message'
```