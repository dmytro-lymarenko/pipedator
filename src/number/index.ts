import { test } from '../test';

export function number<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'number', message || 'Value should be a number');
}
