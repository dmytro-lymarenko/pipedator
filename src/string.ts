import { createValidator } from './core';

export function string() {
	return createValidator({
		message: 'Value should be a string',
		validate: (value, ctx) => (typeof value === 'string' ? null : { value, message: ctx.message, path: ctx.path }),
	});
}

export function empty() {
	return createValidator({
		message: 'Value should be empty',
		validate: (value, ctx) => (value.length === 0 ? null : { value, message: ctx.message, path: ctx.path }),
	});
}
