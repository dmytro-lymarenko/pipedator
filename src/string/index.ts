import { test } from '../test';

export function string<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'string', message || 'Value should be a string');
}
