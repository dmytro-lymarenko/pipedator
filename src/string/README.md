- **Description**

Creates a validator to check if a value is a string or not.

- **How to import**

```typescript
import { string } from 'pipedator';
// or
import { string } from 'pipedator/lib/string';

```
- **Signature**

```typescript
function string<ValidValue = any>(message?: string): Validator<ValidValue>;
```
- **Parameters**

  - `message` - (optional) custom message.


- **Usage**

```typescript
string().validate('Some text'); // valid
string().validate(5); // invalid
string('Custom message').validate(null); // invalid with 'Custom message'
```