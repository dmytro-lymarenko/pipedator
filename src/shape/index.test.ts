import { shape } from './index';
import { number } from '../number';
import { string } from '../string';
import { nullable } from '../nullable';

describe('shape()', () => {
	it('should validate', () => {
		expect(shape({ a: number(), b: string() }).validate({ a: 1, b: 'text' })).toBeNull();
		expect(shape({ a: number(), b: string() }).validate({ a: 1, b: 'text', c: {} })).toBeNull();
	});

	it('should fail', () => {
		expect(shape({ a: number(), b: string() }).validate({ a: '1', b: 'text' })).toMatchSnapshot();
		expect(shape({ a: number(), b: string() }).validate({ a: 1 })).toMatchSnapshot();
	});

	it('should fail and return all errors', () => {
		expect(shape({ a: number(), b: string() }).validate({ a: '1' })).toMatchSnapshot();
	});

	it('should fail and return only the first error', () => {
		expect(shape({ a: number(), b: string() }, { onlyFirstRequirement: true }).validate({ a: '1' })).toMatchSnapshot();
	});

	it('should support custom message', () => {
		expect(shape({ a: number(), b: string() }, undefined, 'Custom message').validate({ a: 1 })).toMatchSnapshot();
	});

	it('should support wrapping validators', () => {
		expect(
			shape({ a: number(), b: string() }, { wrapValidators: validator => nullable(validator) }).validate({ a: null, b: 'text' })
		).toBeNull();
	});
});
