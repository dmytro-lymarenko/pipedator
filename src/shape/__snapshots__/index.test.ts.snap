// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shape() should fail 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a number",
      "path": "a",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`shape() should fail 2`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a string",
      "path": "b",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`shape() should fail and return all errors 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a number",
      "path": "a",
    },
    Object {
      "children": null,
      "message": "Value should be a string",
      "path": "b",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`shape() should fail and return only the first error 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a number",
      "path": "a",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;

exports[`shape() should support custom message 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be a string",
      "path": "b",
    },
  ],
  "message": "Custom message",
  "path": "",
}
`;
